test/e2e/framework/timer/timer.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package timer
+
+import (
+	"time"
+
+	"bytes"
+	"fmt"
+
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/perftype"
+	"sync"
+)
+
+var now = time.Now
+
+// Represents a phase of a test. Phases can overlap.
+type Phase struct {
+	sequenceNumber int
+	name           string
+	startTime      time.Time
+	endTime        time.Time
+}
+
+func (phase *Phase) ended() bool {
+	return !phase.endTime.IsZero()
+}
+
+// End marks the phase as ended, unless it had already been ended before.
+func (phase *Phase) End() {
+	if !phase.ended() {
+		phase.endTime = now()
+	}
+}
+
+func (phase *Phase) label() string {
+	return fmt.Sprintf("%03d-%s", phase.sequenceNumber, phase.name)
+}
+
+func (phase *Phase) duration() time.Duration {
+	endTime := phase.endTime
+	if !phase.ended() {
+		endTime = now()
+	}
+	return endTime.Sub(phase.startTime)
+}
+
+func (phase *Phase) humanReadable() string {
+	if phase.ended() {
+		return fmt.Sprintf("Phase %s: %v\n", phase.label(), phase.duration())
+	} else {
+		return fmt.Sprintf("Phase %s: %v so far\n", phase.label(), phase.duration())
+	}
+}
+
+// A TestPhaseTimer groups phases and provides a way to export their measurements as JSON or human-readable text.
+// It is safe to use concurrently.
+type TestPhaseTimer struct {
+	prep     Phase
+	explicit []*Phase
+	mutex    sync.Mutex
+}
+
+// NewTestPhaseTimer creates a new TestPhaseTimer and starts an implicit "prep" phase.
+func NewTestPhaseTimer() *TestPhaseTimer {
+	return &TestPhaseTimer{prep: Phase{sequenceNumber: 0, name: "prep", startTime: now()}}
+}
+
+// StartPhase starts a new explicit phase.
+// It ends the implicit "prep" phase if it was still running.
+// sequenceNumber is an integer prepended to phaseName in the output, such that lexicographic sorting
+// of phases in perfdash reconstructs the order of execution. Unfortunately it needs to be
+// provided manually, since a simple incrementing counter would have the effect that inserting
+// a new phase would renumber subsequent phases, breaking the continuity of historical records.
+func (timer *TestPhaseTimer) StartPhase(sequenceNumber int, phaseName string) *Phase {
+	timer.prep.End()
+	newPhase := &Phase{sequenceNumber: sequenceNumber, name: phaseName, startTime: now()}
+	timer.mutex.Lock()
```

Comment:
Please move this Lock() statement to the top. Seems like there can be a race with prep.End().

Reply:
I'm not sure I can see the race. `End()` is safe to call concurrently, no?

Reply:
It is modifying the variable `phase.endTime`. And it is possible that two End() calls are at that point in code at the same time.

Reply:
But they, if they are here at the same time, it does not really matter which one of them sets the value. That's what I meant by "safe".
OTOH, this might not be obvious to a casual reader, so I'll move the lock to lower the cognitive burden a little.

Reply:
Yes, I agree with the cognitive burden point too and was about to point it.